File: csharp/Extractjson.cs
// Extract structured invoice data from a Factur-X PDF as JSON using the
// InvoiceXML REST API. Useful for feeding Factur-X invoices into REST
// pipelines, ERPs, or any system that prefers JSON over XML.
//
// Get your API key at: https://www.invoicexml.com/account/authentication
// Full API reference:  https://www.invoicexml.com/docs/api/extract/json
//
// Required NuGet package:
//   dotnet add package Flurl.Http

using Flurl.Http;

public static class ExtractJson
{
    private const string Endpoint = "https://api.invoicexml.com/v1/extract/json";

    public static async Task<string> RunAsync(
        string pdfPath = "invoice.pdf",
        string outputPath = "invoice.json")
    {
        // Get an InvoiceXML API key at: https://www.invoicexml.com/account/authentication
        var apiKey = "INVOICEXML_API_KEY";

        if (!File.Exists(pdfPath))
            throw new FileNotFoundException($"Input PDF not found: {pdfPath}");

        try
        {
            // POST the PDF and receive a JSON document containing the parsed
            // EN 16931 invoice fields (seller, buyer, lines, tax breakdown, totals).
            // Response schema:
            // https://www.invoicexml.com/docs/api/extract/json
            var json = await Endpoint
                .WithOAuthBearerToken(apiKey)
                .PostMultipartAsync(mp => mp
                    .AddFile("file", pdfPath, "application/pdf"))
                .ReceiveString();

            await File.WriteAllTextAsync(outputPath, json);
            Console.WriteLine($"Factur-X data extracted to {outputPath}");
            Console.WriteLine(json);
            return json;
        }
        catch (FlurlHttpException ex)
        {
            var body = await ex.GetResponseStringAsync();
            Console.Error.WriteLine($"InvoiceXML API error {(int?)ex.StatusCode}: {body}");
            throw;
        }
    }
}

File: nodejs/extract-json.js
// Extract Factur-X invoice data as JSON using the InvoiceXML API.
// Get API key: https://www.invoicexml.com/account/authentication
// Docs:        https://www.invoicexml.com/docs/api/extract/json
//
// Requires Node.js 18+. No dependencies.

const fs = require('fs');

const apiKey = 'YOUR_API_KEY';

(async () => {
    const form = new FormData();
    form.append('file', new Blob([fs.readFileSync('invoice.pdf')], { type: 'application/pdf' }), 'invoice.pdf');

    const response = await fetch('https://api.invoicexml.com/v1/extract/json', {
        method: 'POST',
        headers: { Authorization: `Bearer ${apiKey}` },
        body: form,
    });

    const json = await response.text();
    if (!response.ok) {
        console.error(`InvoiceXML API error ${response.status}: ${json}`);
        process.exit(1);
    }

    fs.writeFileSync('invoice.json', json);
    console.log(json);
})();
